src/planner.c

```diff
@@ -312,8 +331,45 @@ timescaledb_planner(Query *parse, int cursor_opts, ParamListInfo bound_params)
 
 	PG_TRY();
 	{
+		PreprocessQueryContext context = { 0 };
+		context.rootquery = parse;
 		if (ts_extension_is_loaded())
-			preprocess_query((Node *) parse, parse);
+		{
+			preprocess_query((Node *) parse, &context);
+
+			/*
+			 * Determine which type of fetcher to use. If set by GUC, use what
+			 * is set. If the GUC says 'auto', use the row-by-row fetcher if we
+			 * have only one distributed table in the query. This enables
+			 * parallel plans on data nodes, which speeds up the query.
+			 * We can't use parallel plans with the cursor fetcher, because the
+			 * cursors don't support parallel execution. This is because a
+			 * cursor can be suspended at any time, then some arbitrary user
+			 * code can be executed, and then the cursor is resumed. The
+			 * parallel infrastructure doesn't have enough reentrability to
+			 * survive this.
+			 * We have to use a cursor fetcher when we have multiple distributed
+			 * tables, because we might first have to get some rows from one
+			 * table and then from another, without running either of them to
+			 * completion first. This happens e.g. when doing a join. If we had
+			 * a connection per table, we could avoid this requirement.
+			 */
+			if (ts_guc_remote_data_fetcher == AutoFetcherType)
+			{
+				if (context.num_distributed_tables >= 2)
+				{
+					data_node_scan_fetcher_type = CursorFetcherType;
+				}
+				else
+				{
```

Comment:
We can have a case when there are Zero distributed hypertables, no?

Reply:
True, I fixed the comment.